File: backend/src/internal/adapters/userStorage/userStorage.ts
import {Logger} from "../../../pkg/logger";
import {QueryFailedError, Repository} from "typeorm";
import {UserDB} from "../../db/postgres/user.models";
import {User} from "../../domain/models/user";
import {unknownBedaDatabase} from "../../domain/services/utils";
import {Beda} from "../../../pkg/beda/Beda";
import {CodeError, Exceptions} from "../../domain/exceptions/exceptions";
import {PG_UNIQUE_VIOLATION} from "@drdgvhbh/postgres-error-codes";
import {DatabaseCodes} from "../../domain/exceptions/database";

export class UserStorage {
    private logger: Logger;
    private userRepo: Repository<UserDB>;

    public constructor(logger: Logger, userRepo: Repository<UserDB>) {
        this.logger = logger;
        this.userRepo = userRepo;
    }

    public async getById(id: string): Promise<User | null> {
        let user: UserDB | null;
        try {
            user = await this.userRepo.findOne({
                where: {
                    id: id,
                },
            });
        } catch (e) {
            throw unknownBedaDatabase()
        }

        if (!user) {
            return null
        }

        return {
            id: user.id,
            username: user.username,
            first_name: user.first_name,
            second_name: user.second_name,
            avatar_url: user.avatar_url,
            day_of_birth: user.day_of_birth,
            gender: user.gender,
        }
    }

    public async updateById(id: string, set: Omit<User, "id">): Promise<void> {
        try {
            await this.userRepo
                .createQueryBuilder('user')
                .update()
                .set({
                    username: set.username,
                    first_name: set.first_name,
                    second_name: set.second_name,
                    avatar_url: set.avatar_url,
                    day_of_birth: set.day_of_birth,
                    gender: set.gender,
                    date_modify: new Date(),
                })
                .where('id = :id', { id: id })
                .execute();
        } catch (e) {
            this.parseUpdateUserError(e)
        }
    }

    private parseUpdateUserError(e: unknown) {
        const beda = new Beda(Exceptions.Database, CodeError.Database)
        if (e instanceof QueryFailedError) {
            const err: any = e
            switch (err.code) {
                case PG_UNIQUE_VIOLATION: {
                    switch (err.constraint) {
                        case 'users_username_uniq':
                            beda.addDesc(DatabaseCodes.UsernameAlreadyExist)
                            break
                        default:
                            beda.addDesc(DatabaseCodes.SomeAlreadyExist)
                            break
                    }
                    break
                }
                default:
                    beda.addDesc(DatabaseCodes.Unknown)
            }
            throw beda
        }
        throw beda
    }
}

File: backend/internal/adapters/userStorage/userStorage.go
package userStorage

import (
	"authService/internal/domain/exceptions"
	"authService/internal/domain/models"
	"authService/pkg/beda"
	"authService/pkg/database/postgres"
	"context"
)

type UserStorage struct {
	client postgres.Client
}

func NewUserStorage(client postgres.Client) *UserStorage {
	return &UserStorage{
		client: client,
	}
}

func (storage *UserStorage) GetById(ctx context.Context, id string) (models.User, error) {
	query := `select (id, email, password_hash, salt, first_name, second_name, day_of_birth, gender, date_create, date_modify) 
from public.user
where id = $1
limit 1`
	rows, err := storage.client.Query(ctx, query, id)
	if err != nil {
		return models.User{}, beda.New(exceptions.ErrorDatabase, err)
	}
	defer rows.Close()

	if !rows.Next() {
		return models.User{}, beda.New(exceptions.ErrorNotFound, nil)
	}

	user := UserDB{}
	err = rows.Scan(&user)
	if err != nil {
		return models.User{}, beda.New(exceptions.ErrorDatabase, err)
	}

	return models.User{
		IdUser:       user.Id,
		Email:        user.Email,
		PasswordHash: user.PasswordHash,
		Salt:         user.Salt,
		FirstName:    user.FirstName,
		SecondName:   user.SecondName,
		DateOfBirth:  user.DayOfBirth,
		Gender:       user.Gender,
		DateCreate:   user.DateCreate,
		DateModify:   user.DateModify,
	}, nil
}

func (storage *UserStorage) GetByEmail(ctx context.Context, email string) (models.User, error) {
	query := `select (id, email, password_hash, salt, first_name, second_name, day_of_birth, gender, date_create, date_modify) 
from public.user
where email = $1
limit 1`
	rows, err := storage.client.Query(ctx, query, email)
	if err != nil {
		return models.User{}, beda.New(exceptions.ErrorDatabase, err)
	}
	defer rows.Close()

	if !rows.Next() {
		return models.User{}, beda.New(exceptions.ErrorNotFound, nil)
	}

	user := UserDB{}
	err = rows.Scan(&user)
	if err != nil {
		return models.User{}, beda.New(exceptions.ErrorDatabase, err)
	}

	return models.User{
		IdUser:       user.Id,
		Email:        user.Email,
		PasswordHash: user.PasswordHash,
		Salt:         user.Salt,
		FirstName:    user.FirstName,
		SecondName:   user.SecondName,
		DateOfBirth:  user.DayOfBirth,
		Gender:       user.Gender,
		DateCreate:   user.DateCreate,
		DateModify:   user.DateModify,
	}, nil
}

func (storage *UserStorage) Create(ctx context.Context, user models.User) error {
	query := `insert into public.user(id, email, password_hash, salt, first_name, second_name, day_of_birth, gender, date_create, date_modify)
	values ($1, $2, $3,$4,$5,$6,$7,$8,$9,$10)`
	row, err := storage.client.Query(
		ctx,
		query,
		user.IdUser,
		user.Email,
		user.PasswordHash,
		user.Salt,
		user.FirstName,
		user.SecondName,
		user.DateOfBirth,
		user.Gender,
		user.DateCreate,
		user.DateModify,
	)
	if err != nil {
		return beda.New(exceptions.ErrorDatabase, err)
	}
	defer row.Close()

	return nil
}
